nixos: systemd-coredump: improve disabled state

squalus opened the pull request:
###### Description of changes

The systemd-coredump module required systemd to be built with `withCoredump=true`, even if the module was disabled.

- allow systemd to be missing systemd-coredump if the module is disabled
- switch to mkDefault for the sysctl config to allow user overrides when the module is disabled
- add nixos tests for both the enabled and disabled cases

Should not trigger any rebuilds.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Mic92 commented on nixos/tests/systemd-coredump-disabled.nix:
```diff
@@ -0,0 +1,21 @@
+
+  nodes.machine = { pkgs, lib, ... }: {
+    systemd.coredump.enable = false;
+    nixpkgs.config.packageOverrides = pkgs: {
+      systemd = pkgs.systemd.override {
+        withCoredump = false;
+      };
+    };
```
This makes the test cheaper:

```suggestion
    systemd.package = pkgs.systemd.override {
        withCoredump = false;
    };
```

systemd still triggers quite a few rebuilds.

squalus replied:
That is indeed better. Updated.

pennae commented on nixos/tests/systemd-coredump-disabled.nix:
```diff
@@ -0,0 +1,19 @@
+  };
+
+  nodes.machine = { pkgs, lib, ... }: {
+    systemd.coredump.enable = false;
+    systemd.package = pkgs.systemd.override {
+      withCoredump = false;
+    };
```
might it make sense to add `withCoredump` to systemd passthru and default `systemd.coredump.enable` to `systemd.package.withCoredump`?

flokli replied:
I assume the PR intent is mostly to not fail ungracefully in case `systemd.package` is overridden with a variant that doesn't include coredump functionality.

Peeking at `systemd.package` passthru to set options sounds like the wrong direction.

squalus replied:
Agreed with @flotkli. I would rather the decision of how to handle coredumps was separate from whether systemd-coredump was built.

flokli commented on nixos/tests/systemd-coredump.nix:
```diff
@@ -0,0 +1,16 @@
+  testScript = ''
+    machine.wait_for_unit("multi-user.target")
+
+    machine.wait_for_unit("systemd-coredump.socket")
```
Can we make this a bit more sophisticated? Try to run a small binary that coredumps, and ensure it's visible when invoking `coredumpctl`?

squalus replied:
Sure. I'll give that a shot.

squalus replied:
Improved the test to verify that a crashing service actually shows up in coredumpctl.

squalus commented:
- Added a test to verify that a crashing service appears in coredumpctl.

flokli commented on nixos/tests/systemd-coredump-disabled.nix:
```diff
@@ -0,0 +1,19 @@
+  };
+
+  testScript = ''
+    machine.wait_for_unit("multi-user.target")
```
In the systemd-coredump disabled, I think we should also let the binary crash here and check for the default behaviour without the systemd sysctl snippet (core dump created in working directory).

You might want to add this logic into systemd-coredump, and override the config with the `specialisation` test feature.

squalus replied:
For the first point, that makes sense. I can add a test to check that the crashing program produces a coredump in the working directory, which is what we expect from the default sysctl config.

I don't understand the second point. Could you elaborate on this?

flokli replied:
If both tests run the binary, expect it to crash and then do some (slightly different) assertions, there's a lot of duplicate code. 

My proposal was to join the two files into a single `.nix` file, and use the `specialisation` test feature (grep `nixos/tests` for some other examples) to create two different VMs - one with coredump enabled, one with it being disabled (and `systemd.package` overridden to a version without support).

Then in the test script, boot both, assert `coredumpctl` showing the dump, and the other VM having produced the coredump in the working dir.

squalus replied:
Ok, that makes sense. I reworked it to use one test file and a specialisation for the disabled case. But I'm having trouble getting it to switch to a new configuration and then reboot. I think I need to run `nix-env` to set the system profile to the new specialisation, and then run `switch-to-configuration boot`, then restart the machine. Haven't been able to get this working yet.

flokli replied:
I think the sysctl's should also get applied on a simple switch, without a reboot.

Check https://github.com/NixOS/nixpkgs/blob/master/nixos/tests/specialisation.nix for a simple example. It should be sufficient to do a `switch-to-configuration test` (which applies, but doesn't update bootloaders), without a reboot.

squalus replied:
The `switch-to-configuration test` operation never finishes. I have the intermediate work here: https://github.com/NixOS/nixpkgs/compare/master...squalus:nixpkgs:coredump-specialisation?expand=1

If I type the `switch-to-configuration test` command manually in the qemu vm, it works fine. But it hangs forever when invoked by the test script.

```
machine: must succeed: /run/current-system/specialisation/disabled/bin/switch-to-configuration test
machine # [    9.303393] nixos[912]: switching to system configuration /nix/store/y5hllpqa87mi4pjnx83lidwgm8kaj1pk-nixos-system-machine-22.11pre-git
machine # stopping the following units: audit.service, backdoor.service, logrotate-checkconf.service, mount-pstore.service, network-local-commands.service, network-setup.service, nscd.service, resolvconf.service, systemd-coredump.socket, systemd-modules-load.service, systemd-sysctl.service, systemd-udevd-control.socket, systemd-udevd-kernel.socket, systemd-udevd.service, systemd-update-done.service
machine # [    9.328545] systemd[1]: Stopping Kernel Auditing...
machine # [    9.337855] systemd[1]: Stopping backdoor.service...
machine # [    9.342305] systemd[1]: Stopped target Local File Systems.
machine # [    9.347503] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[929]: No rules
machine # [    9.354716] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: enabled 0
machine # [    9.357227] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: failure 1
machine # [    9.359783] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: pid 0
machine # [    9.362970] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: rate_limit 0
machine # [    9.364078] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: backlog_limit 64
machine # [    9.365135] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: lost 0
machine # [    9.366064] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: backlog 0
machine # [    9.367003] gs5qwm35p5xdnwgm9v5k6zmlrg6gwg41-audit-stop[932]: backlog_wait_time 60000
machine # [    9.368215] systemd[1]: audit.service: Deactivated successfully.
machine # [    9.369132] systemd[1]: Stopped Kernel Auditing.
machine # [    9.369992] systemd[1]: backdoor.service: Deactivated successfully.
machine # [    9.370990] systemd[1]: Stopped backdoor.service.
```

flokli replied:
Huh, that's odd. @grahamc, any idea on that?

squalus commented:
Added a test for the disabled case. Now we ensure that there's a coredump created in the current working directory of a crashing service, since this is what the default sysctl config is supposed to do.

flokli commented:
@squalus I couldn't get the specialisation to work either, but based on `squalus/coredump-specialisation` I refactored the test to use two VMs with slightly different configs:

```
From de6188983eedefed1200cff1282ae000bc6b7531 Mon Sep 17 00:00:00 2001
From: Florian Klink <flokli@flokli.de>
Date: Mon, 1 Aug 2022 14:00:57 +0700
Subject: [PATCH] nixosTests.systemd-coredump: resort to two machines

It seems the specialisation feature doesn't work.

Build two closures and run the tests with two different VMs for the time
being.
---
 nixos/tests/systemd-coredump.nix | 43 ++++++++++++++++----------------
 1 file changed, 21 insertions(+), 22 deletions(-)

diff --git a/nixos/tests/systemd-coredump.nix b/nixos/tests/systemd-coredump.nix
index e18bb0c1b72..24d6653b4c7 100644
--- a/nixos/tests/systemd-coredump.nix
+++ b/nixos/tests/systemd-coredump.nix
@@ -4,6 +4,15 @@ let
 
   crasher = pkgs.writeCBin "crasher" "int main;";
 
+  commonConfig = {
+    systemd.services.crasher.serviceConfig = {
+      ExecStart = "${crasher}/bin/crasher";
+      StateDirectory = "crasher";
+      WorkingDirectory = "%S/crasher";
+      Restart = "no";
+    };
+  };
+
 in
 
 {
@@ -12,34 +21,24 @@ in
     maintainers = [ squalus ];
   };
 
-  nodes.machine = { pkgs, lib, ... }: {
-    systemd.services.crasher.serviceConfig = {
-      ExecStart = "${crasher}/bin/crasher";
-      StateDirectory = "crasher";
-      WorkingDirectory = "%S/crasher";
-      Restart = "no";
-    };
-
-    specialisation.disabled.configuration.systemd = {
-      coredump.enable = false;
-      package = pkgs.systemd.override {
-        withCoredump = false;
-      };
+  nodes.machine1 = { pkgs, lib, ... }: commonConfig;
+  nodes.machine2 = { pkgs, lib, ... }: commonConfig // {
+    systemd.coredump.enable = false;
+    systemd.package = pkgs.systemd.override {
+      withCoredump = false;
     };
-
   };
 
   testScript = ''
     with subtest("systemd-coredump enabled"):
-      machine.wait_for_unit("multi-user.target")
-      machine.wait_for_unit("systemd-coredump.socket")
-      machine.systemctl("start crasher");
-      machine.wait_until_succeeds("coredumpctl list | grep crasher", timeout=10)
-      machine.fail("stat /var/lib/crasher/core")
+      machine1.wait_for_unit("multi-user.target")
+      machine1.wait_for_unit("systemd-coredump.socket")
+      machine1.systemctl("start crasher");
+      machine1.wait_until_succeeds("coredumpctl list | grep crasher", timeout=10)
+      machine1.fail("stat /var/lib/crasher/core")
 
     with subtest("systemd-coredump disabled"):
-      machine.succeed("/run/current-system/specialisation/disabled/bin/switch-to-configuration test")
-      machine.systemctl("start crasher");
-      machine.wait_until_succeeds("stat /var/lib/crasher/core", timeout=10)
+      machine2.systemctl("start crasher");
+      machine2.wait_until_succeeds("stat /var/lib/crasher/core", timeout=10)
   '';
 })
-- 
2.37.1
```

Maybe we can merge it like this until we figure out how to get the specialisation stuff to work?

squalus committed:
nixos: systemd-coredump: improve disabled state

The systemd-coredump module required systemd to be built with
withCoredump=true, even if the module was disabled.

- allow systemd to be missing systemd-coredump if the module is disabled
- switch to mkDefault for the sysctl config to allow user overrides when
  the module is disabled
- add nixos tests for both the enabled and disabled cases

squalus commented:
> Maybe we can merge it like this until we figure out how to get the specialisation stuff to work?

Sounds good. I updated the PR with your suggestion, slightly modified to get the attrSets combined as expected. (Needed to use recursiveUpdate instead of `//`)

flokli commented:
Thanks for your patience!
